internal/pkg/vm/internal/vmcontext/testing.go

```diff
@@ -302,6 +301,72 @@ func (a *ValidationApplier) ApplyMessage(context *vtypes.ExecutionContext, state
 	return receipt, nil
 }
 
+func toOurBlockMessageInfoType(theirs []vtypes.BlockMessagesInfo) []interpreter.BlockMessagesInfo {
+	ours := make([]interpreter.BlockMessagesInfo, len(theirs))
+	for i, bm := range theirs {
+		ours[i].Miner = bm.Miner
+		for _, blsMsg := range bm.BLSMessages {
+			ourbls := &types.UnsignedMessage{
+				To:         blsMsg.To,
+				From:       blsMsg.From,
+				CallSeqNum: uint64(blsMsg.CallSeqNum),
+				Value:      blsMsg.Value,
+				Method:     blsMsg.Method,
+				Params:     blsMsg.Params,
+				GasPrice:   blsMsg.GasPrice,
+				GasLimit:   types.GasUnits(blsMsg.GasLimit),
+			}
+			ours[i].BLSMessages = append(ours[i].BLSMessages, ourbls)
+		}
+		for _, secpMsg := range bm.SECPMessages {
+			oursecp := &types.SignedMessage{
+				Message: types.UnsignedMessage{
+					To:         secpMsg.Message.To,
+					From:       secpMsg.Message.From,
+					CallSeqNum: uint64(secpMsg.Message.CallSeqNum),
+					Value:      secpMsg.Message.Value,
+					Method:     secpMsg.Message.Method,
+					Params:     secpMsg.Message.Params,
+					GasPrice:   secpMsg.Message.GasPrice,
+					GasLimit:   types.GasUnits(secpMsg.Message.GasLimit),
+				},
+				Signature: secpMsg.Signature,
+			}
+			ours[i].SECPMessages = append(ours[i].SECPMessages, oursecp)
+		}
+	}
+	return ours
+}
+
+func (a *ValidationApplier) ApplyTipSetMessages(state vstate.VMWrapper, blocks []vtypes.BlockMessagesInfo, epoch abi.ChainEpoch, rnd vstate.RandomnessSource) ([]vtypes.MessageReceipt, error) {
+	st := state.(*ValidationVMWrapper)
+
+	// XXX: unsure if this is redundant
+	st.vm.currentEpoch = epoch
+
+	ourBlkMsgs := toOurBlockMessageInfoType(blocks)
+	receipts, err := st.vm.ApplyTipSetMessages(ourBlkMsgs, epoch, rnd)
+	if err != nil {
+		return nil, err
+	}
+
+	// XXX: unsure if this is redundant.
```

Comment:
Please do figure this out before landing. If you can't work it out, I guess leave the code (but not the XXX) here